File: features/person/src/main/java/com/nivasafinance/features/person/service/impl/PersonWriteServiceImpl.java
package com.nivasafinance.features.person.service.impl;

import com.nivasafinance.common.dto.AddressData;
import com.nivasafinance.common.dto.AddressRequest;
import com.nivasafinance.common.dto.IdentifierData;
import com.nivasafinance.common.dto.IdentifierRequest;
import com.nivasafinance.common.exception.BadRequestException;
import com.nivasafinance.features.address.service.AddressDataService;
import com.nivasafinance.features.person.dto.PersonCreateRequest;
import com.nivasafinance.features.person.dto.PersonCreateResponse;
import com.nivasafinance.features.person.dto.PersonUpdateRequest;
import com.nivasafinance.features.person.entity.MobileNumberDetails;
import com.nivasafinance.features.person.entity.Person;
import com.nivasafinance.features.person.exception.PersonExceptionFactory;
import com.nivasafinance.features.person.repository.PersonRepositoryWrapper;
import com.nivasafinance.features.identifier.service.IdentifierService;
import com.nivasafinance.features.person.service.PersonWriteService;
import lombok.AllArgsConstructor;
import org.springframework.context.MessageSource;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

@Service
@AllArgsConstructor
@Transactional
public class PersonWriteServiceImpl implements PersonWriteService {

    private final PersonRepositoryWrapper personRepositoryWrapper;
    private final MessageSource messageSource;
    private final AddressDataService addressDataService;
    private final IdentifierService identifierService;

    @Override
    public PersonCreateResponse createPerson(PersonCreateRequest request) {
        // Extract primary mobile number
        String primaryMobile = extractPrimaryMobileNumber(request.getMobileNumbers());
        
        // Check if primary mobile already exists
        if (primaryMobile != null) {
            Optional<Person> existingPerson = personRepositoryWrapper.findByPrimaryMobileNumber(primaryMobile);
            if (existingPerson.isPresent()) {
                throw PersonExceptionFactory.primaryMobileAlreadyExists(primaryMobile, messageSource);
            }
        }

        // Create Person entity
        Person person = new Person();
        person.setFirstName(request.getFirstName());
        person.setMiddleName(request.getMiddleName());
        person.setLastName(request.getLastName());
        person.setMobileNumbers(request.getMobileNumbers());
        person.setDateOfBirth(request.getDateOfBirth());
        person.setGender(request.getGender());
        
        // Generate display name
        person.setDisplayName(generateDisplayName(
                request.getFirstName(),
                request.getMiddleName(),
                request.getLastName()
        ));

        // Save person
        Person savedPerson = personRepositoryWrapper.saveWithException(person);
        
        return PersonCreateResponse.builder()
                .id(savedPerson.getId())
                .build();
    }

    @Override
    public void updatePerson(Long personId, PersonUpdateRequest request) {
        // Fetch existing person
        Person person = personRepositoryWrapper.findByIdWithException(personId);
        
        // Extract primary mobile number from request
        String primaryMobile = extractPrimaryMobileNumber(request.getMobileNumbers());
        
        // Check if primary mobile exists for a different person
        if (primaryMobile != null) {
            Optional<Person> existingPerson = personRepositoryWrapper.findByPrimaryMobileNumber(primaryMobile);
            if (existingPerson.isPresent() && !existingPerson.get().getId().equals(personId)) {
                throw PersonExceptionFactory.primaryMobileAlreadyExists(primaryMobile, messageSource);
            }
        }

        // Update all person fields
        person.setFirstName(request.getFirstName());
        person.setMiddleName(request.getMiddleName());
        person.setLastName(request.getLastName());
        person.setMobileNumbers(request.getMobileNumbers());
        person.setDateOfBirth(request.getDateOfBirth());
        person.setGender(request.getGender());
        
        // Generate display name
        person.setDisplayName(generateDisplayName(
                request.getFirstName(),
                request.getMiddleName(),
                request.getLastName()
        ));

        // Save person
        personRepositoryWrapper.saveWithException(person);
    }

    private String extractPrimaryMobileNumber(List<MobileNumberDetails> mobileNumbers) {
        if (mobileNumbers == null || mobileNumbers.isEmpty()) {
            return null;
        }
        
        return mobileNumbers.stream()
                .filter(m -> m.getIsPrimary() != null && m.getIsPrimary())
                .map(MobileNumberDetails::getNumber)
                .findFirst()
                .orElseThrow(()-> new BadRequestException("One Mobile Number is required to be marked as primary"));
    }

    private String generateDisplayName(String firstName, String middleName, String lastName) {
        StringBuilder displayName = new StringBuilder();
        
        if (firstName != null && !firstName.trim().isEmpty()) {
            displayName.append(firstName.trim());
        }

        if (middleName != null && !middleName.trim().isEmpty()) {
            if (!displayName.isEmpty()) displayName.append(" ");
            displayName.append(middleName.trim().charAt(0));
        }
        
        if (lastName != null && !lastName.trim().isEmpty()) {
            if (!displayName.isEmpty()) displayName.append(" ");
            displayName.append(lastName.trim());
        }
        
        return !displayName.isEmpty() ? displayName.toString() : null;
    }

    @Override
    public String addAddress(Long personId, AddressRequest request) {

        if (request.getAddressType() == null) {
            throw PersonExceptionFactory.addressTypeMandatory(messageSource);
        } 

        Person person = personRepositoryWrapper.findByIdWithException(personId);
        List<AddressData> addresses = getAddresses(person);

        AddressData addressData = buildAddressData(request, null);
        // Ensure ID is set before adding to list
        if (addressData.getId() == null) {
            addressData.setId(UUID.randomUUID().toString());
        }
        addresses.add(addressData);

        saveAddresses(person, addresses);

        return addressData.getId();
    }

    @Override
    public AddressData updateAddress(Long personId, String addressId, AddressRequest request) {


        if (request.getAddressType() == null) {
            throw PersonExceptionFactory.addressTypeMandatory(messageSource);
        } 

        Person person = personRepositoryWrapper.findByIdWithException(personId);
        List<AddressData> addresses = getAddresses(person);

        AddressData existing = addresses.stream()
                .filter(address -> addressId.equals(address.getId()))
                .findFirst()
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,
                        "Address not found for person"));

        AddressData updatedAddress = buildAddressData(request, existing.getId());

        int index = addresses.indexOf(existing);
        addresses.set(index, updatedAddress);

        saveAddresses(person, addresses);

        return updatedAddress;
    }

    private List<AddressData> getAddresses(Person person) {
        List<AddressData> addresses = person.getAddress();
        return addresses == null ? new ArrayList<>() : new ArrayList<>(addresses);
    }

    private void saveAddresses(Person person, List<AddressData> addresses) {
        if (addresses == null || addresses.isEmpty()) {
            person.setAddress(null);
        } else {
            person.setAddress(new ArrayList<>(addresses));
        }
        personRepositoryWrapper.saveWithException(person);
    }

    private AddressData buildAddressData(AddressRequest request, String addressId) {
        AddressData addressData = addressDataService.createAddressData(request);
        addressData.setId(addressId != null ? addressId : UUID.randomUUID().toString());
        addressData.setAddressType(request.getAddressType());
        return addressData;
    }

    @Override
    public IdentifierData addIdentifier(Long personId, IdentifierRequest request) {
        Person person = personRepositoryWrapper.findByIdWithException(personId);
        List<IdentifierData> identifiers = getIdentifiers(person);

        IdentifierData identifierData = identifierService.createIdentifierData(request);
        identifiers.add(identifierData);

        saveIdentifiers(person, identifiers);

        return identifierData;
    }

    @Override
    public void updateIdentifier(Long personId, UUID identifierId, IdentifierRequest request) {
        Person person = personRepositoryWrapper.findByIdWithException(personId);
        List<IdentifierData> identifiers = getIdentifiers(person);

        IdentifierData existing = identifiers.stream()
                .filter(identifier -> identifierId.equals(identifier.getId()))
                .findFirst()
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,
                        "Identifier not found for person"));

        IdentifierData updatedIdentifier = identifierService.createIdentifierData(request);
        updatedIdentifier.setId(existing.getId());

        int index = identifiers.indexOf(existing);
        identifiers.set(index, updatedIdentifier);

        saveIdentifiers(person, identifiers);

    }

    @Override
    public void deleteIdentifier(Long personId, UUID identifierId) {
        Person person = personRepositoryWrapper.findByIdWithException(personId);
        List<IdentifierData> identifiers = getIdentifiers(person);

        boolean removed = identifiers.removeIf(identifier -> identifierId.equals(identifier.getId()));
        if (!removed) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Identifier not found for person");
        }

        saveIdentifiers(person, identifiers);
    }

    private List<IdentifierData> getIdentifiers(Person person) {
        List<IdentifierData> identifiers = person.getIdentifiers();
        return identifiers == null ? new ArrayList<>() : identifiers;
    }

    private void saveIdentifiers(Person person, List<IdentifierData> identifiers) {
        if (identifiers == null || identifiers.isEmpty()) {
            person.setIdentifiers(null);
        } else {
            person.setIdentifiers(new ArrayList<>(identifiers));
        }
        personRepositoryWrapper.saveWithException(person);
    }

    @Override
    public void updateCreditBureauFields(Long personId, List<Long> cbEnquiryIds, Person.CreditBureauDetails cbDetails) {
        Person person = personRepositoryWrapper.findByIdWithException(personId);

        if (cbEnquiryIds != null) {
            person.setCbEnquiryId(cbEnquiryIds);
        }

        if (cbDetails != null) {
            person.setCbDetails(cbDetails);
        }

        personRepositoryWrapper.saveWithException(person);
    }
}

File: features/person/src/main/java/com/nivasafinance/features/person/service/impl/PersonWriteServiceImpl.kt
package com.nivasafinance.features.person.service.impl

import base.BaseNavigatorService
import com.nivasafinance.features.address.dto.AddressCreateRequest
import com.nivasafinance.features.address.dto.AddressUpdateRequest
import com.nivasafinance.features.address.enum.AddressType
import com.nivasafinance.features.address.exception.AddressTypeAlreadyExistsException
import com.nivasafinance.features.address.service.AddressService
import com.nivasafinance.features.person.dto.EmploymentDetailsCreateRequest
import com.nivasafinance.features.person.dto.EmploymentDetailsResponse
import com.nivasafinance.features.person.dto.EmploymentDetailsUpdateRequest
import com.nivasafinance.features.person.dto.PersonAddressMappingRequest
import com.nivasafinance.features.person.dto.PersonAddressMappingResponse
import com.nivasafinance.features.person.dto.PersonAddressMappingUpdateRequest
import com.nivasafinance.features.person.dto.PersonCreateRequest
import com.nivasafinance.features.person.dto.PersonIdentifierCreateRequest
import com.nivasafinance.features.person.dto.PersonIdentifierResponse
import com.nivasafinance.features.person.dto.PersonIdentifierUpdateRequest
import com.nivasafinance.features.person.dto.PersonUpdateRequest
import com.nivasafinance.features.person.entity.EmploymentDetails
import com.nivasafinance.features.person.entity.Person
import com.nivasafinance.features.person.entity.PersonAddressMapping
import com.nivasafinance.features.person.entity.PersonIdentifier
import com.nivasafinance.features.person.enum.IdentifierType
import com.nivasafinance.features.person.exception.DuplicateIdentifierTypeException
import com.nivasafinance.features.person.exception.DuplicatePrimaryMobileNumberException
import com.nivasafinance.features.person.exception.EmploymentDetailsAlreadyExistsException
import com.nivasafinance.features.person.exception.EmploymentDetailsNotFoundException
import com.nivasafinance.features.person.exception.InvalidAddressTypeException
import com.nivasafinance.features.person.exception.InvalidIdentifierTypeException
import com.nivasafinance.features.person.exception.InvalidMobileNumberException
import com.nivasafinance.features.person.exception.PersonAddressMappingNotFoundException
import com.nivasafinance.features.person.exception.PersonIdentifierNotFoundException
import com.nivasafinance.features.person.exception.PersonNotFoundException
import com.nivasafinance.features.person.exception.PrimaryMobileNumberAlreadyExistsException
import com.nivasafinance.features.person.repository.EmploymentDetailsRepository
import com.nivasafinance.features.person.repository.PersonAddressMappingRepository
import com.nivasafinance.features.person.repository.PersonIdentifierRepository
import com.nivasafinance.features.person.repository.PersonRepository
import com.nivasafinance.features.person.service.PersonWriteService
import com.nivasafinance.features.person.util.MobileNumberValidator
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Service
class PersonWriteServiceImpl(
    private val personRepository: PersonRepository,
    private val personAddressMappingRepository: PersonAddressMappingRepository,
    private val personIdentifierRepository: PersonIdentifierRepository,
    private val employmentDetailsRepository: EmploymentDetailsRepository,
    private val addressService: AddressService
) : PersonWriteService, BaseNavigatorService() {

    @Transactional
    override fun createPerson(request: PersonCreateRequest): UUID {
        request.mobileNumbers.forEach { mobileNumber ->
            if (!MobileNumberValidator.isValidIndianMobileNumber(mobileNumber.number)) {
                throw InvalidMobileNumberException(messageSource)
            }
        }

        val primaryNumbers = request.mobileNumbers.filter { it.isPrimary == true }
        if (primaryNumbers.size > 1) {
            throw DuplicatePrimaryMobileNumberException(messageSource)
        }
        primaryNumbers.forEach { primaryMobile ->
            primaryMobile.number?.let { mobileNumber ->
                val existingPerson = personRepository.findByPrimaryMobileNo(mobileNumber)
                if (existingPerson != null) {
                    throw PrimaryMobileNumberAlreadyExistsException(messageSource)
                }
            }
        }
        val cleanedMobileNumbers = request.mobileNumbers.map { mobileNumber ->
            mobileNumber.copy(number = MobileNumberValidator.cleanMobileNumber(mobileNumber.number))
        }

        val personEntity = Person(
            firstName = request.firstName,
            middleName = request.middleName,
            lastName = request.lastName,
            mobileNumbers = cleanedMobileNumbers,
            email = request.email,
            dateOfBirth = request.dateOfBirth,
            gender = request.gender
        )
        val savedPersonEntity = personRepository.save(personEntity)
        return savedPersonEntity.id!!
    }

    @Transactional
    override fun deletePerson(id: UUID) {
        if (!personRepository.existsById(id)) {
            throw PersonNotFoundException(id, messageSource)
        }
        personRepository.deleteById(id)
    }

    @Transactional
    override fun updatePerson(id: UUID, request: PersonUpdateRequest) {
        val existingPerson = personRepository.findById(id)
            .orElseThrow { PersonNotFoundException(id, messageSource) }

        // Update fields if provided
        request.firstName?.let { existingPerson.firstName = it }
        request.middleName?.let { existingPerson.middleName = it }
        request.lastName?.let { existingPerson.lastName = it }
        request.email?.let { existingPerson.email = it }
        request.dateOfBirth?.let { existingPerson.dateOfBirth = it }
        request.gender?.let { existingPerson.gender = it }

        // Update mobile numbers if provided
        if (request.mobileNumbers.isNotEmpty()) {
            // Validate all mobile numbers format
            request.mobileNumbers.forEach { mobileNumber ->
                if (!MobileNumberValidator.isValidIndianMobileNumber(mobileNumber.number)) {
                    throw InvalidMobileNumberException(messageSource)
                }
            }

            // Validate that there's only one primary mobile number in the request
            val primaryNumbers = request.mobileNumbers.filter { it.isPrimary == true }
            if (primaryNumbers.size > 1) {
                throw DuplicatePrimaryMobileNumberException(messageSource)
            }

            // Check if any of the primary mobile numbers already exist in the database (excluding current person)
            primaryNumbers.forEach { primaryMobile ->
                primaryMobile.number?.let { mobileNumber ->
                    val existingPersonWithMobile = personRepository.findByPrimaryMobileNo(mobileNumber)
                    if (existingPersonWithMobile != null && existingPersonWithMobile.id != id) {
                        throw PrimaryMobileNumberAlreadyExistsException(messageSource)
                    }
                }
            }

            // Clean and format mobile numbers
            val cleanedMobileNumbers = request.mobileNumbers.map { mobileNumber ->
                mobileNumber.copy(number = MobileNumberValidator.cleanMobileNumber(mobileNumber.number))
            }
            existingPerson.mobileNumbers = cleanedMobileNumbers
        }

        personRepository.save(existingPerson)
    }

    // Address mapping operations
    @Transactional
    override fun addAddressToPerson(
        personId: UUID,
        request: PersonAddressMappingRequest
    ): PersonAddressMappingResponse {
        if (!personRepository.existsById(personId)) {
            throw PersonNotFoundException(personId, messageSource)
        }

        // Validate address type
        val addressType = request.addressType
        if (addressType.isNullOrBlank()) {
            throw InvalidAddressTypeException.mandatory(messageSource)
        }

        if (!AddressType.isValid(addressType)) {
            throw InvalidAddressTypeException.withValidTypes(addressType, AddressType.getValidTypes(), messageSource)
        }

        // Check if person already has an address of this type
        if (personAddressMappingRepository.existsByPersonIdAndAddressType(personId, addressType)) {
            throw AddressTypeAlreadyExistsException(personId, addressType, messageSource)
        }

        // Create address from request details
        val addressCreateRequest = AddressCreateRequest(
            addressOne = request.addressOne,
            addressTwo = request.addressTwo,
            landmark = request.landmark,
            district = request.district,
            state = request.state,
            pincode = request.pincode,
            addressSource = request.addressSource ?: "CUSTOMER"
        )

        val addressResponse = addressService.createAddress(addressCreateRequest)
        val addressId = addressResponse.id!!

        val mapping = PersonAddressMapping(
            personId = personId,
            addressId = addressId,
            addressType = addressType
        )

        val savedMapping = personAddressMappingRepository.save(mapping)

        return PersonAddressMappingResponse(
            id = savedMapping.id!!,
            personId = personId,
            address = addressResponse,
            addressType = savedMapping.addressType
        )
    }

    @Transactional
    override fun updatePersonAddressMapping(
        personId: UUID,
        addressId: UUID,
        request: PersonAddressMappingUpdateRequest
    ): PersonAddressMappingResponse {
        if (!personRepository.existsById(personId)) {
            throw PersonNotFoundException(personId, messageSource)
        }

        val mapping = personAddressMappingRepository.findByPersonIdAndAddressId(personId, addressId)
            ?: throw PersonAddressMappingNotFoundException(personId, addressId, messageSource)

        // Validate address type if provided
        request.addressType?.let { addressType ->
            // Validate enum value
            if (!AddressType.isValid(addressType)) {
                throw InvalidAddressTypeException.withValidTypes(
                    addressType,
                    AddressType.getValidTypes(),
                    messageSource
                )
            }

            // Check if person already has an address of this type (excluding current address)
            val existingMappingWithType = personAddressMappingRepository.findByPersonId(personId)
                .find { it.addressType == addressType && it.addressId != addressId }

            if (existingMappingWithType != null) {
                throw AddressTypeAlreadyExistsException(personId, addressType, messageSource)
            }
        }

        // Update mapping fields if provided
        request.addressType?.let { mapping.addressType = it }
        val savedMapping = personAddressMappingRepository.save(mapping)

        // Update address details if pincode is provided (required field)
        val addressResponse = if (request.pincode != null) {
            val addressUpdateRequest = AddressUpdateRequest(
                addressOne = request.addressOne,
                addressTwo = request.addressTwo,
                landmark = request.landmark,
                district = request.district,
                state = request.state,
                pincode = request.pincode!!,
                addressSource = request.addressSource
            )
            addressService.updateAddress(addressId, addressUpdateRequest)
        } else {
            addressService.getAddress(addressId)
        }

        return PersonAddressMappingResponse(
            id = savedMapping.id!!,
            personId = personId,
            address = addressResponse,
            addressType = savedMapping.addressType
        )
    }

    @Transactional
    override fun removeAddressFromPerson(personId: UUID, addressId: UUID) {
        if (!personRepository.existsById(personId)) {
            throw PersonNotFoundException(personId, messageSource)
        }

        val mapping = personAddressMappingRepository.findByPersonIdAndAddressId(personId, addressId)
        if (mapping == null) {
            throw PersonAddressMappingNotFoundException(personId, addressId, messageSource)
        }

        // Delete the mapping first
        personAddressMappingRepository.delete(mapping)

        // Then delete the actual address
        addressService.deleteAddress(addressId)
    }

    // Identifier operations
    @Transactional
    override fun createPersonIdentifier(
        personId: UUID,
        request: PersonIdentifierCreateRequest
    ): PersonIdentifierResponse {
        if (!personRepository.existsById(personId)) {
            throw PersonNotFoundException(personId, messageSource)
        }

        // Validate identifier type
        if (!IdentifierType.isValid(request.type.name)) {
            throw InvalidIdentifierTypeException.withValidTypes(request.type.name, messageSource)
        }

        // Check if identifier type already exists for this person
        if (personIdentifierRepository.existsByPersonIdAndType(personId, request.type)) {
            throw DuplicateIdentifierTypeException(personId, request.type, messageSource)
        }

        val identifier = PersonIdentifier(
            personId = personId,
            identifier = request.identifier,
            type = request.type
        )

        val savedIdentifier = personIdentifierRepository.save(identifier)
        return mapIdentifierToResponse(savedIdentifier)
    }

    @Transactional
    override fun updatePersonIdentifier(id: UUID, request: PersonIdentifierUpdateRequest): PersonIdentifierResponse {
        val identifier = personIdentifierRepository.findById(id)
            .orElseThrow { PersonIdentifierNotFoundException(id, messageSource) }

        request.identifier?.let { identifier.identifier = it }
        request.type?.let { newType ->
            // Validate identifier type
            if (!IdentifierType.isValid(newType.name)) {
                throw InvalidIdentifierTypeException.withValidTypes(newType.name, messageSource)
            }

            // Check if the new type already exists for this person (excluding current identifier)
            val existingIdentifier = personIdentifierRepository.findByPersonIdAndType(identifier.personId, newType)
            if (existingIdentifier != null && existingIdentifier.id != id) {
                throw DuplicateIdentifierTypeException(identifier.personId, newType, messageSource)
            }
            identifier.type = newType
        }

        val savedIdentifier = personIdentifierRepository.save(identifier)
        return mapIdentifierToResponse(savedIdentifier)
    }

    @Transactional
    override fun deletePersonIdentifier(id: UUID) {
        if (!personIdentifierRepository.existsById(id)) {
            throw PersonIdentifierNotFoundException(id, messageSource)
        }
        personIdentifierRepository.deleteById(id)
    }

    // Employment details operations
    @Transactional
    override fun createPersonEmploymentDetails(
        personId: UUID,
        request: EmploymentDetailsCreateRequest
    ): EmploymentDetailsResponse {
        if (!personRepository.existsById(personId)) {
            throw PersonNotFoundException(personId, messageSource)
        }

        if (employmentDetailsRepository.existsByPersonId(personId)) {
            throw EmploymentDetailsAlreadyExistsException(personId, messageSource)
        }

        val employmentDetails = EmploymentDetails(
            personId = personId,
            employerName = request.employerName,
            employerType = request.employerType,
            jobTitle = request.jobTitle,
            department = request.department,
            employmentType = request.employmentType,
            location = request.location,
            salary = request.salary,
            documents = request.documents,
            extData = request.extData
        )

        val savedEmploymentDetails = employmentDetailsRepository.save(employmentDetails)
        return EmploymentDetailsResponse.fromEmploymentDetails(savedEmploymentDetails)
    }

    @Transactional
    override fun updatePersonEmploymentDetails(
        personId: UUID,
        request: EmploymentDetailsUpdateRequest
    ): EmploymentDetailsResponse {
        val employmentDetails = employmentDetailsRepository.findByPersonId(personId)
            ?: throw EmploymentDetailsNotFoundException(personId, messageSource)

        request.employerName?.let { employmentDetails.employerName = it }
        request.employerType?.let { employmentDetails.employerType = it }
        request.jobTitle?.let { employmentDetails.jobTitle = it }
        request.department?.let { employmentDetails.department = it }
        request.employmentType?.let { employmentDetails.employmentType = it }
        request.location?.let { employmentDetails.location = it }
        request.salary?.let { employmentDetails.salary = it }
        request.documents?.let { employmentDetails.documents = it }
        request.extData?.let { employmentDetails.extData = it }

        val updatedEmploymentDetails = employmentDetailsRepository.save(employmentDetails)
        return EmploymentDetailsResponse.fromEmploymentDetails(updatedEmploymentDetails)
    }

    @Transactional
    override fun deletePersonEmploymentDetails(personId: UUID) {
        val employmentDetails = employmentDetailsRepository.findByPersonId(personId)
            ?: throw EmploymentDetailsNotFoundException(personId, messageSource)
        employmentDetailsRepository.delete(employmentDetails)
    }

    private fun mapIdentifierToResponse(identifier: PersonIdentifier): PersonIdentifierResponse {
        return PersonIdentifierResponse(
            id = identifier.id!!,
            personId = identifier.personId,
            identifier = identifier.identifier,
            type = identifier.type
        )
    }
}
